### ISSUE TITLE
 `AnchorImage` init will fail with pytorch models with default fp32 precission

### ISSUE THREAD

**Issue by sakoush** [SeldonIO/alibi#499]
NN pytorch models would default to using `torch.float32` if no dtype is specified, e.g. check https://pytorch.org/docs/stable/generated/torch.nn.Linear.html

These models will fail with `AnchorImage` at initialisation.

When `AnchorImage` is initialised, there is logic that tries to detect the output of classification models (if it is class probabilities we then perform ArgMaxTransformation)

As part of the logic, we send a dummy infer request and the [data](https://github.com/SeldonIO/alibi/blob/ab7dbf9b2a153ff49f29a372d5bdeac56dc7284a/alibi/explainers/anchor_image.py#L605) is constructed using `np.zeros` (which [defaults](https://numpy.org/doc/stable/reference/generated/numpy.zeros.html) to `numpy.fp64`), failing the init for pytorch models with default fp32 precision. 

A couple of ideas to sort this out:

- Pass metadata about the inference model dtype during `AnchorImage` init, which will then be used  to construct data with correct dtype.
- Describe the output of the inference model (whether it is a class label or class probabilities) instead of letting `AnchorImage` trying to infer it (and we will not require a dummy call to the inference model during init, which sounds like a good idea). This is also related to https://github.com/SeldonIO/alibi/issues/492

**Comment by sakoush** [comment]
@jklaise in the case of `mlserver` where this issue surfaced, the inference model is behind a predict api endpoint (no access to the full model torch) and while we could query the model metadata (and hopefully the metadata will have dtypes in it) to replicate what you have done in `image = ...`, it feels that the root cause of this issue is that during init of `AnchorImage` there is "dummy" predict call. 
Alternatively we can have the user specifying if `argmax` transformer is required and not having to do the dummy call, which will also cover a similar issue with `IntegratedGradients`?

In anycase `predict_fn` referenced above is not going to be applicable in `mlserver`.

**Comment by jklaise** [comment]
I see, so the needs of `mlserver` are a bit different from the needs of a pure `alibi` user.

I think it's valuable exploring moving away from dummy calls to the underlying model as this is a pattern where we effectively try to guess the type of input the model expects (although the `alibi` interface always expects black-box models to be of type `Callable[[np.ndarray], np.ndarray]` but the `dtype` detail hasn't been addressed up to now). This would effectively mean an `alibi` API change to ask users more info about their model (related: #492).

So there are a few options:
  1. Keep dummy calls as they're useful for inferring the type of model, but expose additional kwarg for `dtype`
  2. Replace dummy calls and instead of trying to infer the type of model expose additional kwargs for the user to specify that information upfront.

Number 1. should be a fairly simple change with no impact on the current API, the decision would need to be made if we switch to `dtype=np.float32` by default. @sakoush would this be enough for `mlserver` to work?
Number 2. is slightly more complex and would involve a change in the API but is potentially valuable to do in the long term.
